numba/np/arraymath.py

```diff
@@ -3674,107 +3674,102 @@ def bincount_impl(a, weights=None, minlength=0):
-            # since there shouldn't be many of them in practice)
-            for i in range(n, 0, -1):
-                if not np.isnan(a[i - 1]):
-                    return i
-            return 0
-
-        if v_last < v:
-            hi = n
-        else:
-            lo = 0
-            hi = hi + 1 if hi < n else n
-
-        while hi > lo:
-            mid = (lo + hi) >> 1
-            if func(a[mid], (v)):
-                # mid is too low => go up
-                lo = mid + 1
+@register_jitable
+def custom_lt(a, b):
+    if np.isnan(b):
+        return not np.isnan(a)
+    return a < b
+
+
+@register_jitable
+def custom_le(a, b):
+    if np.isnan(b):
+        return True
+    return a <= b
+
+
+def _searchsorted(func_1, func_2):
+    # a facsimile of:
+    # https://github.com/numpy/numpy/blob/4f84d719657eb455a35fcdf9e75b83eb1f97024a/numpy/core/src/npysort/binsearch.cpp#L61  # noqa: E501
+
+    def impl(a, v):
+        min_idx = 0
+        max_idx = len(a)
+
+        out = np.empty(v.size, np.intp)
+
+        last_key_val = v.flat[0]
+
+        for i in range(v.size):
+            key_val = v.flat[i]
+
+            if func_1(last_key_val, key_val):
+                max_idx = len(a)
             else:
-                # mid is too high, or is a NaN => go down
-                hi = mid
-        return lo
-    return searchsorted_inner
+                min_idx = 0
+                if max_idx < len(a):
+                    max_idx += 1
+                else:
+                    max_idx = len(a)
 
+            last_key_val = key_val
 
-_lt = less_than
-_le = register_jitable(lambda x, y: x <= y)
-_searchsorted_left = register_jitable(_searchsorted(_lt))
-_searchsorted_right = register_jitable(_searchsorted(_le))
+            while min_idx < max_idx:
+                # to avoid overflow
+                mid_idx = min_idx + ((max_idx - min_idx) >> 1)
+
+                mid_val = a[mid_idx]
+
+                if func_2(mid_val, key_val):
+                    min_idx = mid_idx + 1
+                else:
+                    max_idx = mid_idx
+
+            out[i] = min_idx
+
+        return out.reshape(v.shape)
+
+    return impl
+
+
+_searchsorted_left = register_jitable(
+    _searchsorted(custom_lt, custom_lt)
+)
+_searchsorted_right_pre_np123 = register_jitable(
+    _searchsorted(custom_lt, custom_le)
+)
+_searchsorted_right_np123_on = register_jitable(
+    _searchsorted(custom_le, custom_le)
+)
 
 
 @overload(np.searchsorted)
 def searchsorted(a, v, side='left'):
     side_val = getattr(side, 'literal_value', side)
+
     if side_val == 'left':
-        loop_impl = _searchsorted_left
+        _impl = _searchsorted_left
     elif side_val == 'right':
-        loop_impl = _searchsorted_right
+        # change in behaviour introduced by
+        # https://github.com/numpy/numpy/pull/21867
+        if numpy_version >= (1, 23):
+            _impl = _searchsorted_right_np123_on
+        else:
+            _impl = _searchsorted_right_pre_np123
```

Comment:
As of today, `side` argument needs to be a literal value known at compile-time. This could be changed to be a runtime value:

```python
    if isinstance(v, types.Array):
        def impl(a, v, side='left'):
            if side == 'left':
                return _searchsorted_left(a, v)
            elif side == 'right':
                return _searchsorted_right_np123_on(a, v)
            else:
                raise ValueError('error!')
```

But can be done in a future work as this goes beyond the scope of this PR.

Reply:
Good point - I tried to keep the changeset as small as possible in this PR; could be a good candidate change though.

Reply:
I added a comment re this point